app/src/studio/core.c

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2024 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <zmk/studio/core.h>
+
+ZMK_EVENT_IMPL(zmk_studio_core_lock_state_changed);
+ZMK_EVENT_IMPL(zmk_studio_core_unlock_requested);
+
+static enum zmk_studio_core_lock_state state = IS_ENABLED(CONFIG_ZMK_STUDIO_LOCKING)
+                                                   ? ZMK_STUDIO_CORE_LOCK_STATE_LOCKED
+                                                   : ZMK_STUDIO_CORE_LOCK_STATE_UNLOCKED;
+
+enum zmk_studio_core_lock_state zmk_studio_core_get_lock_state(void) { return state; }
+
+static void set_state(enum zmk_studio_core_lock_state new_state) {
+    if (state == new_state) {
+        return;
+    }
+
+    state = new_state;
+
+    raise_zmk_studio_core_lock_state_changed(
+        (struct zmk_studio_core_lock_state_changed){.state = state});
+}
+
+#if CONFIG_ZMK_STUDIO_LOCK_IDLE_TIMEOUT_SEC > 0
+
+static void core_idle_lock_timeout_cb(struct k_work *work) { zmk_studio_core_lock(); }
+
+K_WORK_DELAYABLE_DEFINE(core_idle_lock_timeout, core_idle_lock_timeout_cb);
+
+void zmk_studio_core_reschedule_lock_timeout() {
+    k_work_reschedule(&core_idle_lock_timeout, K_SECONDS(CONFIG_ZMK_STUDIO_LOCK_IDLE_TIMEOUT_SEC));
+}
+
+#else
+
+void zmk_studio_core_reschedule_lock_timeout() {}
+
+#endif
+
+void zmk_studio_core_unlock() {
+    set_state(ZMK_STUDIO_CORE_LOCK_STATE_UNLOCKED);
+
+    zmk_studio_core_reschedule_lock_timeout();
+}
+
+void zmk_studio_core_lock() { set_state(ZMK_STUDIO_CORE_LOCK_STATE_LOCKED); }
+void zmk_studio_core_initiate_unlock() {
+    set_state(ZMK_STUDIO_CORE_LOCK_STATE_UNLOCKING);
```

Comment:
Is unlocking some asynchronous task that takes some time to complete, or is this handling a request from Studio to the user to unlock the keyboard, which they confirm with the `&studio_unlock` behavior?

If the latter, would `_UNLOCK_REQUESTED` be a better name for this enum than `_UNLOCKING`? Also, should this reset back to locked after a timeout? It looks like it can just get stuck in this state forever if the user never responds. Or is this state even necessary? I don't see it referenced anywhere else, so I'm not sure what a handler of the "unlock requested" event is expected to do.

Reply:
This is a half implemented feature where I could see a keyboard with a display showing a prompt to the user about unlocking, allowing input to respond. In that case, it's an async workflow that's initiated, then in progress, then aborted or completed. I can scrap for now instead of partially including, if you'd prefer.

Reply:
My first concern is that we have support for locking from unlocked after a timeout, but not this middle state. What happens if Studio requests an unlock, then without unlocking on the keyboard, you restart Studio? It looks like the keyboard gets stuck in this "unlocking" state unless CONFIG_ZMK_STUDIO_LOCK_ON_DISCONNECT is set? Might be a bit confusing if so.

My second and probably more concerning concern is that
```c
if (zmk_studio_core_get_lock_state() != ZMK_STUDIO_CORE_LOCK_STATE_LOCKED) {
  // we're unlocked, so allow things
}
```
looks perfectly reasonable at first glance yet is completely wrong. Would separate bools be any better? That does allow for the not very meaningful state of "is unlocked and unlock is requested", but it would make the above mistake impossible.